// tweening/lib/tween.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _easingFuncFromName = require('./easingFuncFromName');

var _easingFuncFromName2 = _interopRequireDefault(_easingFuncFromName);

var _tweenGenerator = require('./tweenGenerator');

var _tweenGenerator2 = _interopRequireDefault(_tweenGenerator);

var _middleware = require('./middleware');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var tween = function tween(_ref) {
  var _ref$duration = _ref.duration;
  var duration = _ref$duration === undefined ? 200 : _ref$duration;
  var complete = _ref.complete;
  var _ref$easing = _ref.easing;
  var easing = _ref$easing === undefined ? 'easeInOutQuad' : _ref$easing;
  var from = _ref.from;
  var _ref$middleware = _ref.middleware;
  var middleware = _ref$middleware === undefined ? [] : _ref$middleware;
  var next = _ref.next;
  var to = _ref.to;

  if (typeof duration !== 'number' || duration <= 0 || duration == Infinity) {
    throw new Error('Duration must be a positive number less than infinity');
  }

  var easingFunc = typeof easing === 'function' ? easing : (0, _easingFuncFromName2.default)(easing);

  var g = (0, _tweenGenerator2.default)({
    duration: duration,
    easing: easingFunc,
    from: middleware.length ? (0, _middleware.middlewareIn)(from, middleware) : from,
    to: middleware.length ? (0, _middleware.middlewareIn)(to, middleware) : to
  });

  var animate = function animate() {
    var _g$next = g.next();

    var done = _g$next.done;
    var value = _g$next.value;


    var v = middleware.length ? (0, _middleware.middlewareOut)(value, middleware) : value;

    if (done) {
      if (typeof next === 'function') {
        next(to);
      }

      if (typeof complete === 'function') {
        complete();
      }
    } else {
      if (typeof next === 'function') {
        next(v);
      }

      window.requestAnimationFrame(animate);
    }
  };

  window.requestAnimationFrame(animate);
};

exports.default = tween;
